### PR TITLE
Add basic warnings

### PR DESCRIPTION
**Pull request by brgr**
Generate Warnings for float and add a regression test where it warns if a float (double) value might be Nan or +/-infinity -- i.e. we have top

### CODE REVIEW

File: tests/regression/56-floats/03-infinity_or_nan.c [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// PARAM: --enable ana.float.interval
+#include <stdio.h>
+
+void main()
+{
+    double data;
+    fscanf(stdin, "%lf", &data);
```
Comment (Dudeldu) [inline tests/regression/56-floats/03-infinity_or_nan.c]: i know, it was me coming up with that originally, however we are not checking for uninitialized-access. Therefore, we should probably leaf this out, and make minimal examples for our cases.

File: src/cdomains/valueDomain.ml [hunk 2]
```diff
@@ -193,7 +193,7 @@ struct
     let rec zero_init_value (t:typ): t =
       match t with
       | TInt (ikind, _) -> `Int (ID.of_int ikind BI.zero)
-      | TFloat (FDouble, _) -> failwith "todo"
+      | TFloat (FDouble, _) -> failwith "todo zero_init_value"
```
Comment (Dudeldu) [inline src/cdomains/valueDomain.ml]: Undo this unrelated change

File: src/cdomains/floatDomain.ml [hunk 3]
```diff
@@ -37,18 +37,20 @@ module type FloatDomainBase = sig
...
-    | _ -> None
+    let normed = match v with
+      | Some (low, high) -> 
+        if Float.is_finite low && Float.is_finite high then 
+          if low > high then failwith "invalid Interval"
+          else v
+        else None
+      | _ -> None
+    in if normed == None then 
+      Messages.warn ~category:Messages.Category.FloatMessage ~tags:[CWE 189; CWE 739] "Could be +/-infinity or Nan";
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: i would add "Float" somewhere inside the message e.g. "float value could be ...". If i got this right, this version also warns as soon as the result (or whenever norm is called elsewhere) of the operation is top. Maybe we should incorporate this into the message as well " result of operation could be ...".
Apart from that, keep in mind, that in the future we want to support two options on warning with top inputs and warning on top outputs (i guess that is the current version).

File: src/cdomains/floatDomain.ml [hunk 4]
```diff
@@ -37,18 +37,20 @@ module type FloatDomainBase = sig
...
-      else None
-    | _ -> None
+    let normed = match v with
+      | Some (low, high) -> 
+        if Float.is_finite low && Float.is_finite high then 
+          if low > high then failwith "invalid Interval"
+          else v
+        else None
+      | _ -> None
+    in if normed == None then
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: If the intention is a comparison with top, use the is_top function instead of manually comparing

File: src/util/options.schema.json [hunk 5]
```diff
@@ -1697,6 +1697,12 @@
           "type": "boolean",
           "default": true
         },
+        "float": {
+          "title": "warn.float",
+          "description": "float warnings",
+          "type": "boolean",
+          "default": true
```
Comment (Dudeldu) [inline src/util/options.schema.json]: i guess our new options should be defaulted to `false`, so the behaviour of old examples does not change

File: tests/regression/56-floats/03-infinity_or_nan.c [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    
+    double result1 = data + 1.0; // WARN: Could be +/-infinity or Nan
```
Comment (Dudeldu) [inline tests/regression/56-floats/03-infinity_or_nan.c:8]: what about some additional examples inside more elaborate constructs
```c
if ((int)(data - 1.)) // WARN: Could be +/-infinity or Nan
    return 0; // `if` requires an actual body, otherwise it is ignored
while ((int)(data - 2.)); // WARN: Could be +/-infinity or Nan
assert ((int)(data - 3.)); // WARN: Could be +/-infinity or Nan
```

File: src/util/messageCategory.ml [hunk 7]
```diff
@@ -21,12 +21,19 @@ type behavior =
 
 type integer = Overflow | DivByZero [@@deriving eq, ord, hash]
 
+(* Note that this is named '_message' because float itself is a built-in type *)
+type float_message =
```
Comment (Dudeldu) [inline src/util/messageCategory.ml]: this type isn't used anywhere. Maybe the `category` should be extended to `FloatMessage of float_message` ?

**Review by Dudeldu — CHANGES_REQUESTED** [review]

File: tests/regression/56-floats/03-infinity_or_nan.c (hunk 6 above)
Comment (brgr) [reply tests/regression/56-floats/03-infinity_or_nan.c:8]: Do they actually give any more information, besides maybe that it's also warning for minus? I personally prefer it to keep it as simple as needed, and in this case we are not checking for any control flow. But I can also change it if you think it's better :)

File: tests/regression/56-floats/03-infinity_or_nan.c (hunk 1 above)
Comment (brgr) [reply tests/regression/56-floats/03-infinity_or_nan.c]: Good point. Done.

File: src/cdomains/valueDomain.ml (hunk 2 above)
Comment (brgr) [reply src/cdomains/valueDomain.ml]: Done

**Review by Dudeldu — DISMISSED** [review]

File: tests/regression/56-floats/03-infinity_or_nan.c [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// PARAM: --enable ana.float.interval
```
Comment (Dudeldu) [inline tests/regression/56-floats/03-infinity_or_nan.c]: you will probably have to enable the now by default disabled float warnings for this test

Comment (FelixKrayer) [reply tests/regression/56-floats/03-infinity_or_nan.c]: I agree. after that it seems good to merge

Comment (brgr) [reply tests/regression/56-floats/03-infinity_or_nan.c]: Great catch, thank you! This is also done now

**Review by FelixKrayer — APPROVED** [review]
⊤

**Review by Dudeldu — APPROVED** [review]
